Prepare for new OEREBlex API version 1.2.0

tschoppi opened the pull request:
With the upcoming release of OEREBlex, targeted towards the end of 2019, the following additions are made to the API XML response:

- New `doctype` "notice" for additional documents that are not legally binding. In German: "weitere Informationen und Hinweise"
- New attribute `language` for every `document`, which is set to the ISO short code corresponding to the language of the document contents (title, files, etc.).

This pull request aims to facilitate the adoption of the new API version as soon as it becomes available. As the new API response is still subject to change, it might be advisable to not immediately merge this pull request but wait until the API 1.2.0 is stable. That way I can update the PR in accordance with the subsequent modifications.

I would like you to review my changes, especially the one concerning the `doc_id` in the list of documents that the parser returns. It may well be that there are downstream implications that render this approach unviable. In that case, please advise how to handle this new case of `document` elements without the `id` attribute.

kdeininger commented:
Thanks for your contribution! I will mark this PR as "work in progress".

It should be possible to handle the `doc_id` for notices this way, as it is not used in `pyramid_oereb`, but I will have another look on it.

tschoppi commented:
> `doc_id = 'notice'` should work fine with `pyramid_oereb`. But there might be another issue:
> 
> https://github.com/openoereb/geolink_formatter/blob/2c633a84d44b86ac49a0d267286c0b958a62a390/geolink_formatter/parser.py#L108-L111
> 
> As you can see at line 111, there is a check to ensure unique documents. […]

My solution should work; The ID field of the notice element still uniquely identifies all notices. In this case, suffixing `notice` to all IDs (e.g. resulting in `42notice`) "namespaces" the ID correctly and allows filtering of duplicates. Mind you, there should be no duplicates in any case, as deduplication is already performed in OEREBlex.

I hope I have adequately addressed your concerns. Please let me know if I have overlooked something.

tschoppi commented:
I have changed the default version to `1.2.0` and will let you know once this version is live.

tschoppi commented:
There will be an additional change with the new API version concerning files: The `title` attribute has been renamed to the more semantic `filename`, and a new `title` attribute is introduced which may contain a title string.

Please have a look if my proposed changes make sense.

tschoppi commented:
FYI: The new GEO-Link API is live on the preview systems. It is scheduled to go live on the production systems on December 19.

tschoppi commented:
> There will be an additional change with the new API version concerning files: The `title` attribute has been renamed to the more semantic `filename`, and a new `title` attribute is introduced which may contain a title string.

This has been reworked slightly: The `title` attribute remains the same as before (i.e. containing the filename), and the new `description` attribute may contain a user-defined title string.

kdeininger commented on geolink_formatter/parser.py:
```diff
@@ -98,14 +101,26 @@ def from_string(self, xml):
+                    # of the file (may be empty).
                     files.append(File(
-                        title=file_el.attrib.get('title'),
+                        title=file_el.attrib.get('description'),
```
This change might break backward compatibility for users currently using the attribute `title` as file name. I would like to keep things simple and use the attributes from the geoLink schema (`title` and `description`).

kdeininger commented on geolink_formatter/entity.py:
```diff
@@ -172,19 +172,21 @@ def cycle(self):
             category (str): The file's category.
             href (str): The URL to access the file.
             title (str): The file's title.
+            filename (str): The file's filename.
```
```suggestion
            description (str): The file's description.
```

kdeininger commented on geolink_formatter/entity.py:
```diff
@@ -172,19 +172,21 @@ def cycle(self):
         self._title = title
         self._href = href
         self._category = category
+        self._filename = filename
```
```suggestion
        self._description = description
```

kdeininger commented on geolink_formatter/entity.py:
```diff
@@ -200,3 +202,8 @@ def href(self):
+    @property
+    def filename(self):
+        """str: The file's filename."""
+        return self._filename
```
Also use `description` here.

tschoppi commented:
I hope the PR now passes muster.

Note that the formatter still prefers usage of `description` (true title) over `title` (filename), which could be considered a breaking change.

kdeininger commented:
Thank you for applying the requested changes!

We will keep the formatter using the description, if available. It is a nice improvement and the formatter is only meant for rendering the results as a HTML fragment. I also think, that we (BL) are the only canton currently using the formatter in our WebGIS. _pyramid_oereb_  directly uses the `Document` and `File` objects and is not affected by this change.

I will merge this PR and create a new release.
